app/javascript/mastodon/reducers/accounts.ts

```diff
@@ -0,0 +1,72 @@
+import { Map as ImmutableMap, fromJS } from 'immutable';
+
+import { ACCOUNT_REVEAL } from 'mastodon/actions/accounts';
+import { ACCOUNT_IMPORT, ACCOUNTS_IMPORT } from 'mastodon/actions/importer';
+import type { Account as StateAccount } from 'mastodon/initial_state';
+
+// TODO(trinitroglycerin): the value of this map is actually a record type,
+// mapping keys of Account to those values in an immutable map.
+type State = ImmutableMap<string, ImmutableMap<string, unknown>>;
+const initialState: State = ImmutableMap();
+
+interface Status {
+  limited: boolean;
+  hidden: boolean;
+}
+
+interface Counters {
+  followers_count: number;
+  following_count: number;
+  statuses_count: number;
+}
+
+// This complicated type is used to allow for the deletion of the properties in Counters from the Account type.
+//
+// The properties in the Counters type are derived values and are tracked in the account_counters slice.
+//
+// TODO(trinitroglycerin): This additionally adds the derived hidden and limited properties, as these are not present
+// in the jsdoc types. However, there are a few instances where React components rely on these properties existing,
+// so it's possible that the hidden and limited properties should be graduated to the jsdoc types.
```

Comment:
Yeah, what I have atm (I'll push it up in a second) is a hack type:

```ts
interface TypeSafeImmutableMapImpl<T extends object> {
  get<K extends keyof T, V extends T[K] | null | undefined = undefined>(
    key: K,
    notSetValue?: V
  ): V;
}

export type TypeSafeImmutableMap<T extends object> =
  TypeSafeImmutableMapImpl<T> & Map<keyof T, unknown>;


// TODO(trinitroglycerin): This is NOT typed correctly for nested collections.
//
// fromJS(T) will take a T and turn any nested collections into List() or Map() as appropriate, however
// the typing of this function implies that arrays and objects are untouched.
//
// This could be fixed using Pick and generics, maybe.
export function intoTypeSafeImmutableMap<T extends object>(
  t: T
): TypeSafeImmutableMap<T> {
  return fromJS(t);
}
```
This provides type safety at the first level - the following compiles (if `state.accounts` were `Immutable.Map<string, unknown`, it would not):

```ts
const getAccountBase = (
  state: RootState,
  id: string
): TypeSafeImmutableMap<Account> | null => {
  return state.accounts.get(id, null);
};
```

However, there are some patterns that are not possible to make type safe. Notably, the use of `updateIn()` in immutable isn't type safe - it's not possible to type this in typescript without assertions:

```ts
const incrementFollowers = (state: State, accountId: string): State => {
  state
    .updateIn([accountId, 'followers_count'], (num) => (num as number) + 1)
    .updateIn([me, 'following_count'], (num) => (num as number) + 1);
};
```

Reply:
with regards to the usage of `updateIn`, I think this is a case where RTK/using POJOs with immer helps a lot.

Reply:
Yes, unfortunately the `*In` methods in Immutable can not be typed at all :(

This is why moving to plain JS objects + Immer is inevitable if we want to be fully typed.

But `Record` allow you to use dot accessors, so `value.getIn('p1', 'p2', 'p3')` can be written at `value.p1.p2.p3`. This does not solve `updateIn`, but helps in many places.